src/gbm/gblinear_model.cc

```diff
@@ -0,0 +1,99 @@
+/*!
+ * Copyright 2019 by Contributors
+ */
+#include <utility>
+
+#include "gblinear_model.h"
+#include "xgboost/json_io.h"
+
+namespace xgboost {
+namespace gbm {
+
+class LinearReader : public JsonReader {
+ public:
+  explicit LinearReader(StringView str) : JsonReader(str) {}
+
+  void ParseWeight(GBLinearModelParam const& param, std::vector<float>* p_weights) {
+    size_t n_weights = (param.num_feature + 1) * param.num_output_group;
+    p_weights->resize(n_weights);
+    auto& weights = *p_weights;
+    GetChar('[');
+    for (size_t i = 0; i < n_weights; ++i) {
+      char* end {nullptr};
+      auto beg = raw_str_.c_str() + cursor_.Pos();
+
+      weights[i] = std::strtof(beg, &end);
+
+      auto len = std::distance(beg, static_cast<char const*>(end));
+      for  (decltype(len) j = 0 ; j < len; ++j) { cursor_.Forward(); }
+
+      if (i != n_weights - 1) {
+        GetChar(',');
+      }
+    }
+    GetChar(']');
+  }
+};
+
+void GBLinearModel::Save(Json* p_out) const {
+  auto& out = *p_out;
+  out["model_param"] = toJson(param);
+  std::string raw;
+  raw.reserve(weight.size() * 4 /*floats*/ + weight.size() /*comma*/ );
+  raw += '[';
+  size_t const n_weights = weight.size();
+  for (size_t i = 0; i < n_weights; ++i) {
+    auto w = weight[i];
+    if (i != n_weights - 1) {
+      raw += std::to_string(w) + ',';
```

Comment:
I am concerned that std::to_string is locale dependent (https://en.cppreference.com/w/cpp/string/basic_string/to_string)
and in some locales (russian or ukrainian) they use ',' as a floating point which may break the output.

What do you think about using a locale independent function (if we can use C++17 we should consider std::to_chars)

Reply:
Also as far as I understood documentation std::to_string just calls fprintf (https://en.cppreference.com/w/cpp/io/c/fprintf) which by default outputs floating point numbers with precision 6 which is not always enough to reconstruct the same number

Reply:
Good point.  I will look into `std::to_chars`, thanks for the suggestion.

Reply:
Implementing `std::{from, to}_chars` for integer is easy, but floating point number is messy.  I think I'm just gonna use `std::locale` to work around it instead.

Reply:
I think worst case we can set locale at the top of conversion.

I am more concern that we should output at least numeric_limit<float>::max_digits10 digits so we can reconstruct the same float and I do not believe that the default does this (I think it outputs only 6 instead of 9 digits)

Reply:
@trivialfis How about using [`std::sprintf`](https://en.cppreference.com/w/cpp/io/c/fprintf)? This function is governed by the locale set in [`std::setlocale()`](https://en.cppreference.com/w/cpp/locale/setlocale), and the standard guarantees that the default C locale is `"C"`.

Reply:
Are a sanity rules, locales are meant for interaction with a human user not for serialization format. serialization format SHOULD use the invariant C locale and the numerical precision SHOULD enable roundtrip. I've already lost hour debugging ML code that was not doing that, let avoid this nightmare ;)

Reply:
This might be useful but I've not done C++ for a few year and might be outdated: https://www.boost.org/doc/libs/1_59_0/libs/locale/doc/html/std_locales.html

Reply:
@trivialfis FYI, fmtlib is becoming part of C++20: http://www.zverovich.net/2019/07/23/std-format-cpp20.html

Reply:
@hcho3 Good to know.

Reply:
BTW, as a general rule what version of a standard are we using for xgboost?
C++14?

Reply:
@trams We adhere to C++11 standard.